opm/upscaling/SteadyStateUpscalerImplicit_impl.hpp

```diff
@@ -151,6 +150,20 @@ namespace Opm
             return dst;
         }
 
+        /// Make a a vector of interleaved water and oil saturations from
+        /// a vector of water saturations.
+        void waterSatToBothSat(const std::vector<double>& sw,
+                               std::vector<double>& sboth)
+        {
+            int num = sw.size();
```

Comment:
size_t

Reply:
And there my button was pressed...! I am very much against using unsigned integers for anything other than bit-fiddling, and sad that the C++ standards committee did not see things the same way. My rationale: it's better to put the "danger area" (where things behave unintuitively) far away from numbers you are likely to use.

However, if you insist, I'll change it. Do you get a compiler warning for the current code?

Reply:
potentially, you will. but it's not really the unsigned vs signed, it's 32 bit vs 64 bit that makes me knee jerk here. but i don't insist as such.. it's not really relevant in most codes.

Reply:
OK, I'll leave the button to you. I agree the 32 bit vs 64 bit issue is annoying. At some point I assumed "when we move to 64 bit computes int will of course be 64 bit" but again, I was not in agreement with the rest of the world...

Reply:
> it's 32 bit vs 64 bit

I think I've said this before, but I'll mention it for new audiences: One of the world's major operators was starting to run into fundamental limitations on the sizes of the models they could even represent (let alone simulate) in their in-house software and traced those limitations back to among other things using a 32-bit (signed) integer (i.e., `int`) for vector sizes and indexes.  Depending on the flow physics and well models, the maximum number of cells that the software would support could be as low as a few tens of million.  This was getting uncomfortable for their high-resolution assets.

Suffice it to say that they started a "next generation" software project; although not **just** because of the size issue.

---

@blattms said:
> Maybe you can use `auto` as a compromise?

That's the best answer.  `vector<>::size()` returns a `vector<>::size_type` which is *usually* `std::size_t` but need not be.

Reply:
I think using `long int` for everything would satisfy everybody... I doubt it would cause any performance loss, but it is a possibility (that could be addressed in specific cases).

Reply:
> I think using `long int` for everything would satisfy everybody...

Not really.  There's no guarantee that `long` has greater bit-width than `int`.  The standard only guarantees that the bit-width of `long` is at least as big as the bit-width of `int`.

Reply:
... but from a practical point of view, it would give us 64-bit integers on all current and thinkable future platforms. Even ...size_type is not guaranteed to be 64-bit, so only practical considerations count!

Reply:
> only practical considerations count!

And that's why we should write `const auto n = x.size()`;

Reply:
A minor amendment to avoid saying wrong things: I have been made aware that on Windows with Microsoft compilers long int is 32 bit. So I'll amend my position: I think using `ptrdiff_t` would satisfy everybody...